e3db-crypto-android/src/main/java/com/tozny/e3db/crypto/CipherManager.java

```diff
@@ -0,0 +1,112 @@
+ * 
+ * Copyright (c) 2018 
+ * 
+ * All rights reserved.
+ * 
+ * e3db-java
+ * 
+ * Created by Lilli Szafranski on 4/9/18.
+ * 
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+
+import android.content.Context;
+import android.os.Build;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.IvParameterSpec;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.security.spec.AlgorithmParameterSpec;
+
+
+class CipherManager {
+
+    interface GetCipher {
+        Cipher getCipher(Context context, String identifier, SecretKey key) throws Throwable;
+    }
+
+    private static void saveInitializationVector(Context context, String fileName, byte[] bytes) throws Throwable {
+        FileOutputStream fos = null;
+
+        try {
+            fos = new FileOutputStream(new File(FileSystemManager.getInitializationVectorFilePath(context, fileName)));
+            fos.write(bytes);
+            fos.flush();
+
+        } finally {
+            if (fos != null) fos.close();
+        }
+    }
+
+    private static byte[] loadInitializationVector(Context context, String fileName) throws Throwable {
+        FileInputStream fis = null;
+        byte[] bytes;
+
+        try {
+            File file = new File(FileSystemManager.getInitializationVectorFilePath(context, fileName));
+            int fileSize = (int) file.length();
+            bytes = new byte[fileSize];
+            fis = new FileInputStream(file);
+            fis.read(bytes, 0, fileSize);
+
+        } finally {
+            if (fis != null) fis.close();
+        }
+
+        return bytes;
+    }
+
+    static void deleteInitializationVector(Context context, String fileName) throws Throwable {
+        if (new File(FileSystemManager.getInitializationVectorFilePath(context, fileName)).exists()) {
+            File file = new File(FileSystemManager.getInitializationVectorFilePath(context, fileName));
+            file.delete();
+        }
+    }
+
+    static class SaveCipherGetter implements GetCipher {
+
+        @Override
+        public Cipher getCipher(Context context, String identifier, SecretKey key) throws Throwable {
+            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+            cipher.init(Cipher.ENCRYPT_MODE, key);
+
+            saveInitializationVector(context, identifier, cipher.getIV());
+
+            return cipher;
+        }
+    }
+
+    static class LoadCipherGetter implements GetCipher {
+        private static int RECC_AUTH_TAG_LEN = 128;
+
+        @Override
+        public Cipher getCipher(Context context, String identifier, SecretKey key) throws Throwable {
+            AlgorithmParameterSpec params;
+
+            if (key.getClass().getSimpleName().equals("AndroidKeyStoreSecretKey") && Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT)
```

Comment:
Is the SDK version a sufficient test? I really don't like checking the class name unless there really isn't any other way.

Reply:
The issue doesn't have to do with the SDK; the issue has to do with the keystore. I agree that reflection probably isn't the most robust, but it was the simplest.